Add inbox notification to surface cart and checkout blocks to select merchants

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR will add the code necessary to show an inbox message in WooCommerce Admin's dashboard letting merchants know about the Cart and Checkout blocks.

To enable us to target only a percentage of users (and to keep the users we _do_ target consistent) I have used a `wp_options` entry to give each merchant a randomly generated "probability" of being chosen, this number is between 0-100. When the code runs, if this option doesn't exist in the database, the random number will be added. We can then use this for percentage-based targeting since only 10% of users will have numbers 0-10, 50% will have 0-50 etc.

I have also included a list of extensions that will preclude the merchant from being shown the notification (based on [the list of known incompatible extensions](https://docs.woocommerce.com/document/cart-checkout-blocks-support-status/#section-7)). We need to include `woocommerce-gutenberg-products-block` here too because we don't want to show the notification to merchants already running the feature plugin. At the moment this is commented out for testing purposes.

<!-- Don't forget to update the title with something descriptive. -->
<!-- If your pull request implements a feature flag, make sure you update [this doc](../docs/blocks/features-and-blocks-behind-a-flag.md) -->

### Screenshots
![image](https://user-images.githubusercontent.com/5656702/132216173-bc5bda5b-aced-41e6-8681-f7f64cc46f12.png)


<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

### How to test the changes in this Pull Request:

1. Ensure your store is set to US or UK (or add your country locale to the `ELIGIBLE_COUNTRIES` array in `InboxNotifications.php`
1. Open the WooCommerce Admin dashboard (`/wp-admin/admin.php?page=wc-admin`)
2. Check `wp_options` table and verify an entry has been made for `wc_blocks_surface_cart_checkout_probability` - if the number is above 10, change it to be below 10. (This ensures we only target 10% of users)
2. If you've got any of the incompatible plugins active, then you should not see the notification in your inbox.
3. Disable any incompatible plugins and refresh the admin page, you should see a notification about the Cart and Checkout.
4. To delete the notification, add this code to the top of `create_surface_cart_checkout_blocks_notification` ` Notes::delete_notes_with_name( self::SURFACE_CART_CHECKOUT_NOTE_NAME );`
5. Re-test with a store in an ineligible location, or with incompatible plugins active.

<!-- If you can, add the appropriate labels -->

opr committed:
Add InboxNotifications class

This will be used to handle displaying the inbox notification to select merchants

opr committed:
Show inbox notifications on load

opr committed:
Update copy in inbox notification

opr committed:
Update URL and percent chance for a user to be chosen

opr committed:
Add apostrophe to notification content

opr committed:
Only register notifications when woocommerce_init is done

opr committed:
Wait for admin to load before adding note

opr committed:
Add further disallowed plugins

opr commented on src/Domain/Bootstrap.php:
```diff
@@ -74,6 +75,15 @@ protected function init() {
 		$this->register_dependencies();
 		$this->register_payment_methods();
 
+		add_action(
+			'admin_init',
+			function() {
+				InboxNotifications::create_surface_cart_checkout_blocks_notification();
+			},
+			10,
+			0
+		);
+
```
Is there a better place to add this?

opr commented on src/InboxNotifications.php:
```diff
@@ -0,0 +1,142 @@
+		'woocommerce-google-analytics-pro',
+		'woocommerce-memberships',
+		'woocommerce-paypal-payments',
+		'woocommerce-points-and-rewards',
```
```suggestion
```

opr replied:
Reminder for me to remove this since Points and Rewards is compatible now :)

opr committed:
Allow notification to be shown when using WC Points and Rewards

opr committed:
Remove timestamp and content data from the note

This isn't necessary and was only added for testing.

opr committed:
Check for existence of note based on name instead of content_data

frontdevde commented on src/InboxNotifications.php:
```diff
@@ -0,0 +1,134 @@
+			return;
+		}
+
+		// Finally, check if the store's generated % chance is below the % of users we want to surface this to.
```
Would it make sense to check this earlier? If we're excluding 90% of users to start with, we might not want to run the previous checks for those users.

Maybe we could consider changing the order of some of the checks i.e. the ones that exclude the most users first. For example, given the current set of countries, at first glance, it seems like that check would limit the set of eligible users more than the limited list of incompatible extensions. Or if we'd want to prevent a potentially unnecessary DB write we could also consider moving the simple read checks in front of the probability check.

frontdevde reviewed (approved):
Tested a host of different scenarios based on the testing instructions and can confirm everything worked as expected.

The inbox notification shows for probabilities below 10 while not showing for probabilities above.

|Probability 8 | Probability 100 |
|-|-|
|<img width="540" alt="Screenshot 2021-09-10 at 17 31 07" src="https://user-images.githubusercontent.com/1562646/132879084-979d2b04-d6dc-401f-a9a1-e223ef216274.png">| <img width="501" alt="Screenshot 2021-09-10 at 17 30 49" src="https://user-images.githubusercontent.com/1562646/132879106-8de6e87d-aca7-4e0e-a562-f2400690fbdd.png">|

The inbox notification shows for a UK/US-based store while not showing for other countries e.g. Germany.

| UK | Germany |
|-|-|
|<img width="540" alt="Screenshot 2021-09-10 at 17 31 07" src="https://user-images.githubusercontent.com/1562646/132880679-4c2e7a29-5c78-4358-9237-bffdc6f446ff.png">|<img width="501" alt="Screenshot 2021-09-10 at 17 30 49" src="https://user-images.githubusercontent.com/1562646/132880707-db7eda11-059d-4c57-86d3-f190ba5fbfbf.png">|

The inbox notification shows when no unsupported extension is active while not showing when one is active.

| No unsupported extensions | mailpoet active |
|-|-|
|<img width="540" alt="Screenshot 2021-09-10 at 17 31 07" src="https://user-images.githubusercontent.com/1562646/132880679-4c2e7a29-5c78-4358-9237-bffdc6f446ff.png">|<img width="501" alt="Screenshot 2021-09-10 at 17 30 49" src="https://user-images.githubusercontent.com/1562646/132880707-db7eda11-059d-4c57-86d3-f190ba5fbfbf.png">|

In addition, I also tested if increasing the percentage of users to target works as expected and can confirm it did. 

One thing I noticed is that if the inbox notice shows once and then you switch to a non-eligible country, activate an incompatible plugin, or change the `wc_blocks_surface_cart_checkout_probability` > 10, the inbox notice still continues to show. That doesn't seem ideal but I assume that's how inbox notifications generally work?

Other than that, LGTM 👍

PS: I really appreciated the detailed code comments that made it easy to follow. Keep it up!👏

opr commented:
> One thing I noticed is that if the inbox notice shows once and then you switch to a non-eligible country, activate an incompatible plugin, or change the wc_blocks_surface_cart_checkout_probability > 10, the inbox notice still continues to show. That doesn't seem ideal but I assume that's how inbox notifications generally work?

According to item 2 in the docs https://github.com/woocommerce/woocommerce-admin/blob/38fbb4aedd0f21cf21ddaa6af93e48564fd35324/docs/examples/activity-panel-inbox.md#inbox-design-constraints-and-best-practices we shouldn't delete notes - How do you feel about breaking this rule in the case that they activate an ineligible plugin? I think keeping the notice after a country change is fine, after all we don't have any actual limits on this anyway.

opr committed:
Create delete note function

We'll be deleting notes in the case that an ineligible plugin becomes active.

opr committed:
Reorder the eligible plugin and % targeting check

This will let us target a truer % of users, and also will let us delete the note if a user activates an ineligible plugin.

opr commented:
Hi @frontdevde thanks for your review!

I have made some changes now:
- Add the `delete_surface_cart_checkout_blocks_notification` which will delete the note.
- Make the "eligible plugin check" the first check to run - if an ineligible plugin is found, then run the delete method. This will cater for the situation where a merchant activates an incompatible plugin. The notice should be deleted in this case.
- Make the percentage check run last, so we can target a more accurate percentage of users who meet the criteria, and also prevent a `wp_options` write if we don't need to.

It would be great if you could take a look at this and give it a 👍🏼  if you agree with the changes!

frontdevde reviewed (approved):
Appreciate you acting on the feedback. Retested with activating/deactivating an unsupported extension and LGTM 👍 

|Before|Activating mailpoet|Deactivating mailpoet|
|-|-|-|
|<img width="535" alt="Screenshot 2021-09-13 at 12 30 04" src="https://user-images.githubusercontent.com/1562646/133069106-e53c7ba6-e200-41cd-9818-40d3984611bd.png">|<img width="505" alt="Screenshot 2021-09-13 at 12 31 42" src="https://user-images.githubusercontent.com/1562646/133069132-b06688c3-32d1-45b2-b87a-419c149732ec.png">|<img width="534" alt="Screenshot 2021-09-13 at 12 32 03" src="https://user-images.githubusercontent.com/1562646/133069159-615b2511-0f43-433f-b5b7-d15c42462e7c.png">|

opr committed:
Remove code which was added for testing purposes

The code to return if the feature plugin was active was enabled, woocommerce-gutenberg-products-block was added as an ineligible plugin.
